src/typechecker.erl

```diff
@@ -2909,116 +3025,154 @@ type_check_function(Env, {function,_, Name, NArgs, Clauses}) ->
             ,constraints:empty()};
         {elem_ty, Tys, Cs} ->
-            lists:foldl(fun ({Pat, TTy}, {Env1, Cs1}) ->
-                                {Env2, Cs2} = add_type_pat(Pat, TTy, TEnv, Env1),
-                                {Env2, constraints:combine(Cs1, Cs2)}
-                        end
-                       ,{VEnv, Cs}
-                       ,lists:zip(Pats, Tys));
+            {PatTys, UBounds, VEnv1, Cs1} = add_types_pats(Pats, Tys, TEnv, VEnv),
+            {type(tuple, PatTys)
+            ,type(tuple, UBounds)
+            ,VEnv1
+            ,constraints:combine(Cs, Cs1)};
         {elem_tys, Tyss, Cs} ->
             %% TODO: This code approximates unions of tuples with tuples of unions
             Unions = lists:map(fun (UnionTys) ->
                                        type(union, UnionTys)
                                end
                               ,transpose(Tyss)),
-            lists:foldl(fun ({Pat, Union}, {Env1, Cs1}) ->
-                                {Env2, Cs2} = add_type_pat(Pat, Union, TEnv, Env1),
-                                {Env2, constraints:combine(Cs1, Cs2)}
-                        end
-                       ,{VEnv, Cs}
-                       ,lists:zip(Pats, Unions));
+            {_PatTys, _UBounds, VEnv1, Cs1} = add_types_pats(Pats, Unions, TEnv, VEnv),
+            {type(none), Ty, VEnv1, constraints:combine(Cs, Cs1)};
         {type_error, _Type} ->
             throw({type_error, pattern, P, Tuple, Ty})
     end;
 add_type_pat(Atom = {atom, P, _}, Ty, TEnv, VEnv) ->
     case subtype(Atom, Ty, TEnv) of
         {true, Cs} ->
-            {VEnv, Cs};
+            {Atom, Atom, VEnv, Cs};
         false ->
             throw({type_error, pattern, P, Atom, Ty})
     end;
 add_type_pat(Nil = {nil, P}, Ty, TEnv, VEnv) ->
-    case subtype({type, P, nil, []}, Ty, TEnv) of
+    case subtype(NilTy = type(nil), Ty, TEnv) of
         {true, Cs} ->
-            {VEnv, Cs};
+            {NilTy, NilTy, VEnv, Cs};
         false ->
             throw({type_error, pattern, P, Nil, Ty})
     end;
 add_type_pat(CONS = {cons, P, PH, PT}, ListTy, TEnv, VEnv) ->
+    %% TODO: Return non-empty list type as upper bound.
     case expect_list_type(normalize(ListTy, TEnv), dont_allow_nil_type) of
         any ->
             VEnv2 = add_any_types_pat(PH, VEnv),
-            add_type_pat(PT, ListTy, TEnv, VEnv2);
+            TailTy = normalize(type(union, [ListTy, type(nil)]), TEnv),
+            {_TailPatTy, _TauUBound, VEnv3, Cs} = add_type_pat(PT, TailTy, TEnv, VEnv2),
+            {type(none), ListTy, VEnv3, Cs};
         {elem_ty, ElemTy, Cs1} ->
-            {VEnv2, Cs2} = add_type_pat(PH, ElemTy, TEnv, VEnv),
-            {VEnv3, Cs3} = add_type_pat(PT, ListTy, TEnv, VEnv2),
-            {VEnv3, constraints:combine([Cs1, Cs2, Cs3])};
+            {_PatTy1, _UBound1, VEnv2, Cs2} = add_type_pat(PH, ElemTy, TEnv, VEnv),
+            TailTy = normalize(type(union, [ListTy, type(nil)]), TEnv),
```

Comment:
This seems wrong for improper lists. Or maybe not, see above that I'm not entirely sure of the improper list semantics.

Reply:
Improper lists aren't that complicated. These improper list types exist (apart from aliases):

* `maybe_improper_list(Type1, Type2)`
* `nonempty_improper_list(Type1, Type2)`
* `nonempty_maybe_improper_list(Type1, Type2)`

There is no (possibly empty) `improper_list(Type1, Type2)` because "empty" doesn't make sense for this one. The tail of a `nonempty_improper_list(Type1, Type2)` is either a `nonempty_improper_list(Type1, Type2)` or a `Type2`.

Reply:
Actually, I didn't consider improper lists. I should exclude them to be on the safe side.

Reply:
I filed #110 for my improper list thoughts.